### PR TITLE
Backgroundvariations

### PR DESCRIPTION
**Pull request by atdotde**
WIP Introduve concurrency in computation of plan variations.

### Describe the pull request:

- [ ] Bug fix
- [x] Functional change
- [ ] New feature
- [ ] Code cleanup
- [ ] Build system change
- [ ] Documentation change
- [ ] Language translation

### Changes made:
1) Prevent infinite deco times by not taking stops longer than 2 days.
2) Do the variations calculation in a separate thread. Use signalling to update planner notes
3) Make 2) a compile time option

### Additional information:
This might have hard to detect (or maybe even not so hard to detect) race conditions. Please help
me identifying those by exercising the planner! @RickMWalsh @sfuchs79 you have done so successfully in the past.

### CODE REVIEW

File: core/planner.c [hunk 1]
```diff
@@ -688,7 +688,6 @@ bool plan(struct deco_state *ds, struct diveplan *diveplan, struct dive *dive, i
 	int decostopcounter = 0;
```
Comment (neolit123) [inline core/planner.c:688]: small typo in the commit message of 772a554c -  `finieshed`.

File: core/planner.c [hunk 2]
```diff
@@ -623,6 +623,9 @@ bool enough_gas(int current_cylinder)
```
Comment (neolit123) [inline core/planner.c:623]: a429ea98
i just asked @sfuchs79 to amend his commit messages where the commit title should be a separate sentence and the commit message body should start on a new sentence. @dirkhh what do you think about commit titles warping into the commit messages like this?

just a nitpick, ultimately.

File: ReleaseNotes/ReleaseNotes.txt [hunk 3]
```diff
@@ -20,6 +20,7 @@ Some of the changes since _Subsurface_ 4.7.4
   of the previous templates would be created (#847)
 - Fix issues with filters not updating after changes to the dive list
   (#551, #675)
+- Compute plan variations in the background to improve UI responsivity
```
Comment (neolit123) [inline ReleaseNotes/ReleaseNotes.txt]: sorry, i just merged something else and that may have created a conflict for `ReleaseNotes.txt`.

**Review by neolit123 — COMMENTED** [review]
made a couple of comments.

File: ReleaseNotes/ReleaseNotes.txt (hunk 3 above)
Comment (neolit123) [reply ReleaseNotes/ReleaseNotes.txt]: posted a topic at the ML about this.

**Comment by sfuchs79** [comment]
@atdotde 
After my first quick tests I found these issues (not all are (fully) related to the multi thread implementation):
- Discrepancy between runtime at beginning of plan and runtime of table of 1min
grafik
- Plan (all results - main plan plus variations) not updated when changing combo boxes quickly - is this same as @RickMWalsh finding?! Screenshots showing two examples.
grafik
grafik
- Finally we also need to talk about the real results of the planner variations ;-)
Either I still do not understand them correctly or they are simply wrong.
BTW: I guess I could change the code easily to do variations for +-3m and +-3min instead of +-1m and +-1min? That may make debugging more easy because the values will be higher...

**Comment by atdotde** [comment]
@sfuchs79 your first point (different total runtimes) you resolved yourself. The others (with the missing updates) should (I hope) be fixed by 9c44a02.

Will comment also on #878.

**Comment by RickMWalsh** [comment]
I've done a little bit of testing and mostly it works well, but it doesn't play nice with recreational mode.

Recreational mode:
In recreational mode, variations are given (sometimes), but don't make sense.  What does "Runtime: 72min + 1085:20 /m + 814:00 /min" mean to the user (or developer)?
Now, plan a long decompression dive (say 100m and 100min) in Buhlmann or VPM-B mode.  Then switch to recreational mode (leave the entered profile way excess of NDLs), and switch back.  Sometimes, but seemingly not always, the variations stop working.  The text VARIATIONS is displayed and nothing else, and it doesn't seem to be recoverable without closing subsurface and starting again.  Changing the profile, switching deco modes, etc. don't fix the problem.
I'm guessing the second part of this issue is a lost thread, but I have no idea how to debug it.  The easy fix, is to disable planner variations when in recreation mode.  I've fixed this in my build and once I remember how to send a pull request on a pull request I will.

**Comment by atdotde** [comment]
After a full morning spent with the debugger, I am much more happy with the performance here. There is a real chance to make it in time for the 4.6.5 release.

**Comment by dirkhh** [comment]
I think you missed the opportunity to ask for a 4.6.5 release. But 4.7.5 might be possible :-)

**Comment by dirkhh** [comment]
The CI failures are odd. Especially the Android build appearing to look for desktop-widget include files. I haven't spent enough time staring at the output, but this doesn't seem right.

**Comment by sfuchs79** [comment]
May it be that it simply has to be:
`#include "desktop-widgets/mainwindow.h"`
instead of
`#include <desktop-widgets/mainwindow.h>`
?

**Comment by atdotde** [comment]
Hi,


> On 1. Dec 2017, at 16:23, Dirk Hohndel <notifications@github.com> wrote:
> 
> The CI failures are odd. Especially the Android build appearing to look for desktop-widget include files. I haven't spent enough time staring at the output, but this doesn't seem right.


I have no clue why Linux and Android fail while Mac and Windows work. Could somebody with easy access to a Linux box please check if this is only a problem with travis or if it is real?

Thanks
Robert

--
.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oOo.oO
Robert C. Helling     Elite Master Course Theoretical and Mathematical Physics
                      Scientific Coordinator
                      Ludwig Maximilians Universitaet Muenchen, Dept. Physik
                      Phone: +49 89 2180-4523  Theresienstr. 39, rm. B339
                      http://www.atdotde.de

Enhance your privacy, use cryptography! My PGP keys have fingerprints
A9D1 A01D 13A5 31FA 6515  BB44 0820 367C 36BC 0C1D    and
DCED 37B6 251C 7861 270D  5613 95C7 9D32 9A8D 9B8F

**Comment by dirkhh** [comment]
Not criticizing you @atdotde - just asking... have you considered adding a Linux VM to test things like this?
I'll pull your branch on my VM to see what's up.

**Comment by sfuchs79** [comment]
@dirkhh 
Here on Ubuntu 16.04 this compiles w/o error.
But next poor guess (because this also compiles here):
Replace in mainwindow.h
`#include "ui_mainwindow.h"
#include "ui_plannerDetails.h"`
with
`#include "desktop-widgets/ui_mainwindow.h"
#include "desktop-widgets/ui_plannerDetails.h"`

**Comment by dirkhh** [comment]
It compiles fine here as well. I need to do this on a different machine where I have what I hope is a fairly exact replica of the Travis environment. But that will be in a couple of hours :-(

**Comment by dirkhh** [comment]
Gotcha. diveplannermodel.cpp now suddenly includes mainwindow.h
That is an absolute NONO. Cannot do. Whatever it is that requires this include is wrong and needs to be done differently.

(sidenote: I saw quite a few "if(" with no space before the '(' - that's also something I'd love to not see)

**Comment by dirkhh** [comment]
Oh the irony - simply REMOVING that line makes no difference here. So just don't include that, period

**Comment by atdotde** [comment]
Adopted your suggestions. Let's see if this solves the problem. Have to run now, will check in again tomorrow morning.

**Comment by atdotde** [comment]
Looks better.

**Comment by dirkhh** [comment]
So tests pass. @sfuchs79 @RickMWalsh are you ok with this being merged?

**Comment by RickMWalsh** [comment]
I won't have a chance to test the last couple of commits from @atdotde for a couple of days I think. But they look good and I'm happy with this getting merged

**Comment by dirkhh** [comment]
OK. The feedback seems to indicate that this is fine, I don't think having it sit another month will really get us to a better result... so I'm merging this.

**Pull request merged**